File: WIP -- chngr/RobotPlayer.scala
package randomplayer

import battlecode.common._
import scala.util.control.Breaks

object RobotPlayer {
  def run(rc : RobotController) {
    while(true) {
      try {
        if(rc.getType() == RobotType.HQ)
          hqCode(rc)
        if(rc.getTyoe() == RobotType.SOLDIER)
          soldierCode(rc)
      }
    }
  }

  // Takes the RobotController for HQ and run code here
  def hqCode(hq : RobotController) {
    var roundNum : Int = Clock.getRoundNum()
    var com : Communicator = new Communicator(roundNum)
    var notNuking : Boolean = com.checkNuke(roundNum)

    // Spawns in the direction of enemy hq
    def regularSpawn() : Unit = {
      var dir : Direction = hq.getLocation.directionTo(hq.senseEnemyHQLocation())
      if(hq.canMove(dir))
        hq.spawn(dir)
    }
    def research(upgrade : Upgrade => Unit) = {
      if(!hq.hasUpgrade(upgrade))
        researchUpgrade(upgrade)
    }

    try{
      // Check nuke reserach progress every 100 rounds
      if(notNuking && roundNum % 100 == 0){
        if(hq.senseEnemyNukeHalfDone()){
          com.nukeWarning()
          notNuking = false
        }
      }
      // Check is the HQ is active --- not already in the process
      if(hq.isActive()){
        // Spawn for now
        regularSpawn()
        hq.yield()
      }
    }
    catch{
      e : Exception => {
        println("caught exception: ")
        e.printStackTrace()
      }
    }
  }

  // Takes the RobotController for SOLDIERS and run code here
  def soldierCode(rc: RobotController) {
    var roundNum : Int = Clock.getRoundNum()
    var curLoc : MapLocation = rc.getLocation()
    var team : Team = rc.getTeam()

    // Tries to move toward home as much as possible; returns BOOL to show success
    def retreat() : Boolean = {
      var homeDir : Direction = curLoc.directionTo(rc.senseHQLocation())
      moveToLocation(homeDir)
    }

    // Tries to move toward location; returns boolean indicating sucess
    def moveToLocation(loc : MapLocation => Boolean) = {
      var locDir : Direction = curLoc.directionTo(loc)
      var succeed : Boolean = true
      val dist = curDir.distanceSquaredTo(loc)
      // Try to move toward direction
      // If not possible to move in one orthogonal direction towards location, don't move
      if(rc.isActive() && dist > 0){
        if(locDir.isDiagonal()){
          for(i <- 1 to 5){
            if(rc.canMove(locDir) && noMines(locDir)){
              rc.move(locDir)
              break
            }
            else{
              if(i % 2 == 0){
                for(_ <- 1 to i)
                  locDir.rotateLeft()
              }
              else if(i == 5)
                succeed = false
              else{
                for(_ <- 1 to i)
                  locDir.rotateRight()
              }
            }
          }
        }
        else{
          for(i <- 1 to 3){
            if(rc.canMove(locDir) && noMines(locDir)){
              rc.move(locDir)
              break
            }
            else{
              if(i == 1)
                locDir.rotateRight()
              else if(i == 2){
                for(_ <- 1 to i)
                  locDir.rotateLeft()
              }
              else
                succeed = false
            }
          }
        }
      }
      succeed
    }

    def noMines(dir : Direction => Boolean) = {
      var loc = curDir.add(dir)
      var senseMine(loc) = mine
      mine == null || mine == team
    }
    try{
      if(rc.isActive()){
        // Checks if on enemy mine
        if(rc.senseMine(curLoc) == team.opponent())
          retreat()
      }
    }
    catch{
      case e: Exception => {
        println("caught exception: ")
        e.printStackTrace()
      }
    }
  }
}

File: teams/javatestplayer/RobotPlayer.java
package javatestplayer;

import battlecode.common.*;

/** The example funcs player is a player meant to demonstrate basic usage of the most common commands.
 * Robots will move around randomly, occasionally mining and writing useless messages.
 * The HQ will spawn soldiers continuously.
 */
public class RobotPlayer {
	public static void run(RobotController rc) {
		while (true) {
			try {
				if (rc.getType() == RobotType.HQ) {
          hqCode(rc);
				} else if(rc.getType() == RobotType.ARTILLERY){
		  artilleryCode(rc);		
				} else if (rc.getType() == RobotType.SOLDIER) {
          Combat fighter = new Combat(rc);

          // Work if the soldier is active
          if(rc.isActive()){
            if(rc.senseNearbyGameObjects(Robot.class, 14, rc.getTeam().opponent()).length > 0)
              fighter.fight();
            else{
              MapLocation goal = rc.senseEnemyHQLocation();

              // Find encampments!
              MapLocation[] nearbyEncampments = rc.senseEncampmentSquares(rc.getLocation(), 32, Team.NEUTRAL);
              if(nearbyEncampments.length != 0){
                goal = nearbyEncampments[nearbyEncampments.length - 1];
                if(rc.senseEncampmentSquare(rc.getLocation())){
                  if(rc.senseCaptureCost() <= rc.getTeamPower()){
                    RobotType build;
                    // ## testing with artillery
                    build = RobotType.ARTILLERY;
//                    if(Math.random() > 0.5)
//                      build = RobotType.SUPPLIER;
//                    else
//                      build = RobotType.GENERATOR;
                    rc.captureEncampment(build);
                  }
                }
              }
              // Move in a random direction
              if(Math.random() < 0.1 && rc.senseMine(rc.getLocation()) == null){
                rc.layMine();
              }
              else if(Math.random() < 0.9){
                if(goal == null){
                  goal = rc.senseEnemyHQLocation();
                }
                Direction dir;
                // Move towards enemy hq
                double randVal = Math.random();
                if(randVal > 0.4){
                  dir = rc.getLocation().directionTo(goal);
                }
                // Else randomly choose direction
                else if(randVal > 0.1){
                  dir = Direction.values()[(int)(Math.random() * 8)];
                }
                else{
                  Robot[] nearbyAllies = rc.senseNearbyGameObjects(Robot.class, rc.getLocation(), 32, rc.getTeam());
                  if(nearbyAllies.length > 0){
                    dir = rc.getLocation().directionTo(rc.senseRobotInfo(nearbyAllies[0]).location);
                  }
                  else{
                    dir = rc.getLocation().directionTo(rc.senseHQLocation());
                  }
                }
                // Check if the thing is a mine
                if(rc.senseMine(rc.getLocation().add(dir)) != null && rc.senseMine(rc.getLocation().add(dir)) != rc.getTeam()){
                  if(Math.random() > 0.1)
                    rc.defuseMine(rc.getLocation().add(dir));
                  else
                    rc.yield();
                }
                else if(rc.canMove(dir)){
                  rc.move(dir);
                }
                else{
                  goal = null;
                }
              }
            }
          }
          rc.yield();
        }
      } catch (Exception e) {
        //e.printStackTrace();
      }
    }
  }
  private static void hqCode(RobotController rc) throws GameActionException{
    if (rc.isActive()) {
      // Spawn a soldier
      Direction dir = rc.getLocation().directionTo(rc.senseEnemyHQLocation());
      if (rc.canMove(dir) && Clock.getRoundNum() % 2 != 0)
        rc.spawn(dir);
      // Tries to spawn, otherwise shall research instead
      else{
        Upgrade[] upgrades = {Upgrade.PICKAXE, Upgrade.FUSION, Upgrade.DEFUSION, Upgrade.VISION, Upgrade.NUKE};
        for(int i = 0; i < 5; i++){
          if(rc.hasUpgrade(upgrades[i]))
            continue;
          else{
            rc.researchUpgrade(upgrades[i]);
            break;
          }
        }
      }
    }
  }
  private static void artilleryCode(RobotController rc) throws GameActionException{
	  if(rc.isActive()){
			// Choose target
			// ## should be like combat, build local map and search for good location
			Robot[] enemies = rc.senseNearbyGameObjects(Robot.class, 64, rc.getTeam().opponent());
			if(enemies.length > 0) rc.attackSquare(rc.senseLocationOf(enemies[0]));
		}
  }
}
